lib/TPP/ConvertTppToIdentity.cpp

```diff
@@ -0,0 +1,219 @@
+              if (haveMatched[i].haveMatch) {
+                sizes.push_back(rewriter.getIndexAttr(
+                    packOp.getStaticInnerTiles()[haveMatched[i].index]));
+              } else {
+                sizes.push_back(rewriter.getIndexAttr(1));
+              }
+            }
+
+            auto tensorExtractType =
+                tensor::ExtractSliceOp::inferCanonicalRankReducedResultType(
+                    packOp.getStaticInnerTiles().size(),
+                    packOp.getSource().getType().cast<RankedTensorType>(),
+                    offsets, sizes, strides);
+
+            auto tensorExtract = rewriter.create<tensor::ExtractSliceOp>(
+                loc, tensorExtractType.cast<RankedTensorType>(),
+                packOp.getSource(), offsets, sizes, strides);
+
+            SmallVector<OpFoldResult> insertSliceOffsets;
+            for (int i = 0; i < numLoops; i++) {
+              int indirection = i;
+              if (packOp.getOuterDimsPerm().size() > 0) {
+                indirection = packOp.getOuterDimsPerm()[i];
+              }
+              insertSliceOffsets.push_back(localIvs[indirection]);
+            }
+            for (size_t i = numLoops; i < packOp.getDestRank(); i++) {
+              insertSliceOffsets.push_back(rewriter.getIndexAttr(0));
+            }
+            SmallVector<OpFoldResult> insertSliceSizes;
+            for (int i = 0; i < numLoops; i++) {
+              insertSliceSizes.push_back(rewriter.getIndexAttr(1));
+            }
+            for (size_t i = numLoops; i < packOp.getDestRank(); i++) {
+              insertSliceSizes.push_back(rewriter.getIndexAttr(
+                  packOp.getStaticInnerTiles()[i - numLoops]));
+            }
+
+            SmallVector<OpFoldResult> insertSliceStrides(
+                packOp.getDestRank(), rewriter.getIndexAttr(1));
+            auto insertSliceOp = rewriter.create<tensor::InsertSliceOp>(
+                loc, tensorExtract.getResult(), iterArgs[0], insertSliceOffsets,
+                insertSliceSizes, insertSliceStrides);
+            return {insertSliceOp};
+          });
+      rewriter.replaceAllUsesWith(packOp.getResult(),
+                                  loopNest.loops[0].getResults()[0]);
+      return success();
+    }
+    return failure();
+  }
+};
+
+void populateTppToIdentityPatterns(RewritePatternSet &patterns) {
+  // clang-format off
+     patterns.add<ConvertTppIdentityOp>(patterns.getContext());
+  // clang-format on
+}
+
+struct ConvertTppToIdentity
+    : public ConvertTppToIdentityBase<ConvertTppToIdentity> {
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    populateTppToIdentityPatterns(patterns);
+    (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
+    return;
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<scf::SCFDialect>();
+    registry.insert<tpp::TppDialect>();
```

Comment:
We're not using TPP identity any more, so I guess we can drop the requirement.

For transpose copies we can keep them as `linalg.transpose` and lower straight to XSMM.

For VNNI work we'll have to work out how we support it in `linalg` or if we lower everything to XSMM at a high level (and make it work on `tensor`?).

Reply:
For the now the goal is to have the two packs needed for VNNI lowered into a single `scf.for` loop nest with 2 memref copy (one inner at the length of VNNI packing into a small temp buffer, one outer which copies the temp buffer into the final destination buffer). This will allow for fusion of packing with GEMM, what we need (hopefully) and much later in the lowering pass we can match the 2 `scf.for` + outer memref copy into on `XSMM` call for norm->vnni. Lowering to XSMM at high-level could kill fusion in tiled formulation.